xonsh/jobs.py

```diff
@@ -97,43 +102,47 @@ def _give_terminal_to(pgid):
     except OSError:
         _shell_tty = None
 
+    def _get_next_task():
+        """ Get the next active task and put it on top of the queue"""
+        selected_task = None
+        for tid in tasks:
+            task = get_task(tid)
+            if not task['bg'] and task['status'] == RUNNING:
+                selected_task = tid
+                break
+        if selected_task == None:
+            return
+        tasks.remove(selected_task)
+        tasks.appendleft(selected_task)
+        return get_task(selected_task)
 
-    def wait_for_active_job(signal_to_send=None):
+
+    def wait_for_active_job():
         """
         Wait for the active job to finish, to be killed by SIGINT, or to be
         suspended by ctrl-z.
         """
+
         _clear_dead_jobs()
-        act = builtins.__xonsh_active_job__
-        if act is None:
-            return
-        job = builtins.__xonsh_all_jobs__[act]
-        obj = job['obj']
-        if job['bg'] and job['status'] == 'running':
+
+        active_task = _get_next_task()
+
+        # Return when there are no foreground active task
+        if active_task is None:
+            _give_terminal_to(_shell_pgrp)  # give terminal back to the shell
             return
-        pgrp = job['pgrp']
 
+        pgrp = active_task['pgrp']
+        obj = active_task['obj']
         # give the terminal over to the fg process
         _give_terminal_to(pgrp)
-        # if necessary, send the specified signal to this process
-        # (this hook  was added because vim, emacs, etc, seem to need to have
-        # the terminal when they receive SIGCONT from the "fg" command)
-        if signal_to_send is not None:
-            if signal_to_send == signal.SIGCONT:
-                job['status'] = 'running'
-
-            os.kill(obj.pid, signal_to_send)
 
-            if job['bg']:
-                _give_terminal_to(_shell_pgrp)
-                return
+        _continue(active_task)
 
         _, wcode = os.waitpid(obj.pid, os.WUNTRACED)
         if os.WIFSTOPPED(wcode):
-            job['bg'] = True
-            job['status'] = 'stopped'
             print()  # get a newline because ^Z will have been printed
-            print_one_job(act)
+            active_task['status'] = STOPPED
```

Comment:
Yeah, I think the performance hit is minimal enough for the gains in user readability.